src/web/lib/useVisibility.ts

```diff
@@ -0,0 +1,30 @@
+import { useEffect, useState, useRef } from 'react';
+
+const useVisibility = (threshold: number) => {
+    const [isVisible, updateVisibility] = useState<boolean>(false);
+    const [node, setNode] = useState<HTMLElement | null>(null);
+
+    const observer = useRef<IntersectionObserver | null>(null);
+
+    useEffect(() => {
+        if (observer.current) observer.current.disconnect();
+
+        observer.current = new window.IntersectionObserver(
+            ([entry]) => updateVisibility(entry.isIntersecting),
+            {
+                root: null,
+                threshold,
+            },
+        );
+
+        const { current: currentObserver } = observer;
+
+        if (node) currentObserver.observe(node);
+
+        return () => currentObserver.disconnect();
+    }, [node, threshold]);
```

Comment:
By the way, I'm so stealing this for DCR

Reply:
I played with using this for DCR, [branch here](https://github.com/guardian/dotcom-rendering/tree/oliver/lazy-hydrate). Thanks for doing the grunt work to prove this approach out!

I made a few changes:

1. I changed `isVisible` to `hasBeenSeen` - eg. don't change state when an entry goes out of the viewport. I think we only have this act once use case so I thought it'd be simpler this way (rather than trying to support both cases with extra props)

2. I made it so we can pass in the full Intersection options object because I wanted to use `rootMargin` instead of `threshold`

3. I removed putting the observer in a ref. Mainly because I couldn't get my head around why it was there but I may well be missing a performance gain by having it so feel free to fix that

It looks like:

```
import { useEffect, useState } from 'react';

export const useVisibility = (options: IntersectionObserverInit) => {
    const [hasBeenSeen, setHasBeenSeen] = useState<boolean>(false);
    const [ref, setRef] = useState<HTMLElement | null>(null);

    useEffect(() => {
        const observer = new IntersectionObserver(([entry]) => {
            if (entry.isIntersecting) setHasBeenSeen(true);
        }, options);        

        if (ref) observer.observe(ref);

        return () => {
            if (ref) {
                observer.disconnect();
            }
        };
    }, [ref, options]);

    return [hasBeenSeen, setRef];
};
```

What do you think?

Reply:
Nice! Brief thoughts below, but would be great to chat IRL if you've got a moment:

> I changed isVisible to hasBeenSeen

That sounds good to me. The reason I didn't do that (and I'm sucking up that work in the caller) was to keep this general, and let the caller decide. But if our use-cases so far are all around seeing something for the first time then maybe that's the way to go.

> I made it so we can pass in the full Intersection options object because I wanted to use rootMargin instead of threshold

:+1:

> I removed putting the observer in a ref.

There's some reasoning around it being in a ref in [the blog post I based this on](https://medium.com/the-non-traditional-developer/how-to-use-an-intersectionobserver-in-a-react-hook-9fb061ac6cb5).

Reply:
Maybe we should pair on finalising the implementation of the hook together and then rebase our respective branches on top?